week_3/project/week_3.py

```diff
@@ -19,28 +18,47 @@
 from project.types import Aggregation, Stock
 
 
-@op
-def get_s3_data():
-    # Use your ops from week 2
-    pass
-
-
-@op
-def process_data():
-    # Use your ops from week 2
-    pass
+@op(
+    out={"stocks": Out(dagster_type=List[Stock])},
+    required_resource_keys={"s3"},
+    config_schema={"s3_key": str},
+    tags={"kind": "s3"},
+    description="Get a list of stocks from an S3 file",
+)
+def get_s3_data(context):
+    stocks = list()
+    key = context.op_config["s3_key"]
+    for row in context.resources.s3.get_data(key):
+        stock = Stock.from_list(row)
+        stocks.append(stock)
+    return stocks
+
+
+@op(
+    ins={"stocks": In(dagster_type=List[Stock])},
+    out={"aggregation": Out(dagster_type=Aggregation)},
+    description="Find the largest stock value",
+)
+def process_data(stocks):
+    big_stock = max(stocks, key=lambda x:x.high)
```

Comment:
This will fail if `stocks` is an empty list, which it can be (dagster will happily pass an empty result on from the previous step).

Reply:
Thanks will fix in week 4! 😃